File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n=len(nums)
        suff = [1]*n
        for i in range(n-2, -1, -1):
            suff[i] = suff[i+1]*nums[i+1]
        curr = 1
        for i in range(n):
            suff[i] *= curr
            curr*=nums[i]
        return suff

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        
        int[] suff = new int[n];
        suff[n-1] = 1;
        
        for(int i=n-2; i>=0; --i)
            suff[i] = suff[i+1] * nums[i+1];
        
        int[] ans = new int[n];
        int cur = 1;
        for(int i=0; i<n; ++i){
            ans[i] = cur * suff[i];
            cur *= nums[i];
        }
        
        return ans;
    }
}